src/target/riscv/riscv.c

```diff
@@ -604,90 +605,146 @@ static int maybe_add_trigger_t1(struct target *target, struct trigger *trigger)
 	return ERROR_OK;
 }
 
-static int maybe_add_trigger_t2(struct target *target, struct trigger *trigger)
+struct trigger_request_info {
+	riscv_reg_t tdata1;
+	riscv_reg_t tdata2;
+	riscv_reg_t imask;
+};
+
+static int try_setup_single_match_trigger(struct target *target,
+		struct trigger *trigger, int trigger_type,
+		struct trigger_request_info trig_info)
 {
-	int ret;
-	riscv_reg_t tdata1, tdata2;
+	int ret = ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+	RISCV_INFO(r);
 
+	for (unsigned int idx = 0;
+			find_next_trigger(target, trigger_type, false, &idx) == ERROR_OK &&
+			ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+			++idx) {
+		ret = set_trigger(target, idx, trig_info.tdata1, trig_info.tdata2,
+				trig_info.imask);
+		if (ret == ERROR_OK) {
+			r->trigger_unique_id[idx] = trigger->unique_id;
+			return ERROR_OK;
+		}
+	}
+	return ret;
+}
+
+static int try_setup_chained_match_triggers(struct target *target,
+		struct trigger *trigger, int trigger_type,
+		struct trigger_request_info t1,
+		struct trigger_request_info t2)
+{
+	int ret = ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
 	RISCV_INFO(r);
 
-	tdata1 = 0;
+	for (unsigned int idx = 0;
+			find_next_trigger(target, trigger_type, true, &idx) == ERROR_OK &&
+			ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+			++idx) {
+		ret = set_trigger(target, idx, t1.tdata1, t1.tdata2, t1.imask);
+		if (ret != ERROR_OK)
+			continue;
+		ret = set_trigger(target, idx + 1, t2.tdata1, t2.tdata2, t2.imask);
+		if (ret == ERROR_OK) {
+			r->trigger_unique_id[idx] = trigger->unique_id;
+			r->trigger_unique_id[idx + 1] = trigger->unique_id;
+			return ERROR_OK;
+		} else
+			/* Undo the setting of the previous trigger */
+			set_trigger(target, idx, 0, 0, 0);
```

Comment:
Detail: If the `checkpatch` script won't complain about the braces `{` `}` here, I'd add them.

```suggestion
		} else {
			/* Undo the setting of the previous trigger */
			set_trigger(target, idx, 0, 0, 0);
		}			
```

Reply:
Addressed. There is actually no need for `else`, because inside the `if` there is a `return`